### PR TITLE
feat: add alter share tenants sql

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

feat: add alter share tenants sql

Fixes https://github.com/datafuselabs/databend/issues/7043

### CODE REVIEW

File: common/ast/tests/it/testdata/statement.txt [hunk 1]
```diff
@@ -6090,3 +6090,93 @@ RevokeShareObject(
 )
 
 
+---------- Input ----------
+ALTER SHARE a ADD TENANTS = b,c;
+---------- Output ---------
+ALTER SHARE a ADD TENANTS  b  ,  c
```
Comment (Xuanwo) [inline common/ast/tests/it/testdata/statement.txt:6096]: display is incorrect.

**Comment by Xuanwo** [comment]
Can you give me an rfc on how these features will in general? A overall picture makes it much easier for me to understand how share works.

So far, I don't how will share work. How will share play with grants, revoke, alter? So I can't decide whether current implementation is good or not.

**Comment by Xuanwo** [comment]
Most important thing to me: the relation between share, tenant, users, databese, catalog, table.

A full example that shows how users will play with share.

**Comment by lichuang** [comment]
> Can you give me an rfc on how these features will in general? A overall picture makes it much easier for me to understand how share works.
> 
> So far, I don't how will share work. How will share play with grants, revoke, alter? So I can't decide whether current implementation is good or not.

https://docs.snowflake.com/en/user-guide/data-sharing-provider.html

File: common/ast/src/ast/statements/share.rs [hunk 2]
```diff
@@ -96,3 +96,37 @@ impl Display for RevokeShareObjectStmt<'_> {
         Ok(())
     }
 }
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AlterShareAccountsStmt<'a> {
```
Comment (Xuanwo) [inline common/ast/src/ast/statements/share.rs:101]: How about changing into `AlterShareTenantsStmt`?

File: common/ast/src/ast/statements/share.rs [hunk 3]
```diff
@@ -96,3 +96,37 @@ impl Display for RevokeShareObjectStmt<'_> {
...
+    pub share: Identifier<'a>,
+    pub if_exists: bool,
+    pub tenants: Vec<Identifier<'a>>,
+    pub add: bool,
```
Comment (Xuanwo) [inline common/ast/src/ast/statements/share.rs:105]: How about changing into `is_add`? Or, we can use an `enum` here which makes it easier for parser.

File: common/ast/src/ast/statements/share.rs [hunk 4]
```diff
@@ -96,3 +96,37 @@ impl Display for RevokeShareObjectStmt<'_> {
...
+            write!(f, "IF EXISTS ")?;
+        }
+        write!(f, "{}", self.share)?;
+        if self.add {
+            write!(f, " ADD TENANTS = ")?;
+        } else {
+            write!(f, " REMOVE TENANTS = ")?;
+        }
+        let mut first = true;
+        for account in self.tenants.iter() {
```
Comment (Xuanwo) [inline common/ast/src/ast/statements/share.rs:121]: Use `self.tenants.iter().map(|v|v.to_string()).join(",")` instead.

File: common/ast/src/parser/statement.rs [hunk 5]
```diff
@@ -886,6 +899,7 @@ pub fn statement(i: Input) -> IResult<StatementMsg> {
             | #drop_share: "`DROP SHARE [IF EXISTS] <share_name>`"
             | #grant_share_object: "`GRANT { USAGE | SELECT | REFERENCE_USAGE } ON { DATABASE db | TABLE db.table } TO SHARE <share_name>`"
             | #revoke_share_object: "`REVOKE { USAGE | SELECT | REFERENCE_USAGE } ON { DATABASE db | TABLE db.table } FROM SHARE <share_name>`"
+            | #alter_share_accounts: "`ALTER SHARE [IF EXISTS] <share_name> {ADD | REMOVE} TENANTS = tenant [, tenant, ...]`"
```
Comment (Xuanwo) [inline common/ast/src/parser/statement.rs:902]:
```suggestion
            | #alter_share_accounts: "`ALTER SHARE [IF EXISTS] <share_name> { ADD | REMOVE } TENANTS = tenant [, tenant, ...]`"
```

File: common/ast/src/ast/statements/share.rs [hunk 6]
```diff
@@ -17,6 +17,7 @@ use std::fmt::Formatter;
 
 use common_meta_app::share::ShareGrantObjectName;
 use common_meta_app::share::ShareGrantObjectPrivilege;
+use itertools::Itertools;
```
Comment (Xuanwo) [inline common/ast/src/ast/statements/share.rs:20]: Why we need `Itertools`?

Comment (lichuang) [reply common/ast/src/ast/statements/share.rs:20]: > Why we need `Itertools`?

line 124:
```
accounts: tenants.iter().map(|v| v.to_string()).collect_vec(),
```

File: query/src/sql/planner/plans/share.rs [hunk 7]
```diff
@@ -105,3 +105,18 @@ impl RevokeShareObjectPlan {
         Arc::new(DataSchema::empty())
     }
 }
+
+// Alter Share Accounts Plan
```
Comment (Xuanwo) [inline query/src/sql/planner/plans/share.rs:109]:
```suggestion
// Alter Share Tenants Plan
```

File: query/src/sql/planner/plans/share.rs [hunk 8]
```diff
@@ -105,3 +105,18 @@ impl RevokeShareObjectPlan {
...
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct AlterShareAccountsPlan {
+    pub share: String,
+    pub if_exists: bool,
+    pub accounts: Vec<String>,
+    pub is_add: bool,
+}
+
+impl AlterShareAccountsPlan {
```
Comment (Xuanwo) [inline query/src/sql/planner/plans/share.rs:118]: ditto

**Review by Xuanwo — APPROVED** [review]

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**